go/kbfs/libkbfs/kbfs_cr_test.go

```diff
@@ -772,6 +774,157 @@ func TestBasicCRFileConflict(t *testing.T) {
+	var userName1, userName2 kbname.NormalizedUsername = "u1", "u2"
+	config1, _, ctx, cancel := kbfsOpsConcurInit(t, userName1, userName2)
+	defer kbfsConcurTestShutdown(t, config1, ctx, cancel)
+
+	config2 := ConfigAsUser(config1, userName2)
+	defer CheckConfigAndShutdown(ctx, t, config2)
+
+	// Enable journaling on user 2
+	tempdir, err := ioutil.TempDir(os.TempDir(), "journal_for_fail_fix")
+	defer os.RemoveAll(tempdir)
+	require.NoError(t, err)
+	err = config2.EnableDiskLimiter(tempdir)
+	require.NoError(t, err)
+	err = config2.EnableJournaling(ctx, tempdir,
+		TLFJournalBackgroundWorkEnabled)
+	require.NoError(t, err)
+	jManager, err := GetJournalManager(config2)
+	require.NoError(t, err)
+	err = jManager.EnableAuto(ctx)
+	require.NoError(t, err)
+
+	clock, _ := newTestClockAndTimeNow()
+	config2.SetClock(clock)
+
+	name := userName1.String() + "," + userName2.String()
+
+	t.Log("User 1 creates a file a/b.")
+	rootNode1 := GetRootNodeOrBust(ctx, t, config1, name, tlf.Private)
+
+	kbfsOps1 := config1.KBFSOps()
+	dirA1, _, err := kbfsOps1.CreateDir(ctx, rootNode1, "a")
+	require.NoError(t, err)
+	fileB1, _, err := kbfsOps1.CreateFile(ctx, dirA1, "b", false, NoExcl)
+	require.NoError(t, err)
+	err = kbfsOps1.SyncAll(ctx, rootNode1.GetFolderBranch())
+	require.NoError(t, err)
+
+	t.Log("User 2 looks up file a/b.")
+	rootNode2 := GetRootNodeOrBust(ctx, t, config2, name, tlf.Private)
+
+	kbfsOps2 := config2.KBFSOps()
+	dirA2, _, err := kbfsOps2.Lookup(ctx, rootNode2, "a")
+	require.NoError(t, err)
+	fileB2, _, err := kbfsOps2.Lookup(ctx, dirA2, "b")
+	require.NoError(t, err)
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	err = SetCRFailureForTesting(ctx, config2, rootNode2.GetFolderBranch(),
+		alwaysFailCR)
+	require.NoError(t, err)
+
+	t.Log("Disable updates on user 2.")
+	c, err := DisableUpdatesForTesting(config2, rootNode2.GetFolderBranch())
+	require.NoError(t, err)
+	err = DisableCRForTesting(config2, rootNode2.GetFolderBranch())
+	require.NoError(t, err)
+
+	t.Log("User 1 writes to file a/b.")
+	data1 := []byte{1, 2, 3, 4, 5}
+	err = kbfsOps1.Write(ctx, fileB1, data1, 0)
+	require.NoError(t, err)
+	err = kbfsOps1.SyncAll(ctx, fileB1.GetFolderBranch())
+	require.NoError(t, err)
+
+	t.Log("User 2 writes to file a/b without having heard user 1's update.")
+	data2 := []byte{5, 4, 3, 2, 1}
+	err = kbfsOps2.Write(ctx, fileB2, data2, 0)
+	require.NoError(t, err)
+	err = kbfsOps2.SyncAll(ctx, fileB2.GetFolderBranch())
+	require.NoError(t, err)
+
+	t.Log("Reenable updates and wait for CR to fail.")
+	c <- struct{}{}
+	err = RestartCRForTesting(
+		libcontext.BackgroundContextWithCancellationDelayer(), config2,
+		rootNode2.GetFolderBranch())
+	require.NoError(t, err)
+
+	t.Log("Try to SyncFromServer on user 2.")
+	ctxWithSoonCancel, _ := context.WithTimeout(ctx, 50*time.Millisecond)
+	err = kbfsOps2.SyncFromServer(ctxWithSoonCancel,
+		rootNode2.GetFolderBranch(), nil)
+	require.Error(t, err, "context deadline exceeded")
```

Comment:
This doesn't do what you intended, I think.  This is just checking that `err` is not `nil`, and then if it is `nil`, it will print "context deadline exceeded" as a message.  If you look at the test output, it's actually failing because a single CR run isn't taking it out of conflict mode as expected.  So you can either check that `err.Error()` contains the expected string "Conflict resolution didn't take us out of staging", or (better) you can make that a typed error and use `require.IsType()` to check for it.

Reply:
It's not getting this error, though - it's getting a context deadline exceeded, since the function called will call CR over and over again until it succeeds.

Reply:
Huh?  Looks like Jenkins passed on the new code that doesn't do this check.

Reply:
But it failed locally for me after I made the change you suggested below.

Reply:
Oh, it's nondeterministic. It sometmes gets one and sometimes the other.

Reply:
I'm gonna extend the timeout by a factor of 2 and just require that there be an error.

Reply:
Really?  The version you have in github (commit 4a11e67c9) succeeds 100x in a row for me.  Maybe you can share logs with me?  I'm not sure how you'd be seeing the deadline exceeded error.  `SyncFromServer` only retries if it got out of CR mode somehow, or was never in it (as was the case before when the callbacks were being nil'd out).

Reply:
Huh, maybe I changed some other line - not able to produce that behavior anymore.